### PR TITLE
Make SVD sparse, 7x speedup, fixes #14

### LINKED ISSUE #14: SVD should be sparse

### PR DESCRIPTION
**Pull request by jilljenn**
svd2.py: 0.226060s
vs
svd.py: 1.590889s
=> 7x speedup because we never compute the big matrix.

### CODE REVIEW

File: zero/svd2.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+from zero.recommendation_algorithm import (RecommendationAlgorithm,
+                                           register_algorithm)
+from scipy.sparse import csr_matrix, diags
+from scipy.sparse.linalg import svds
+
+
+def remove_mean(sp_matrix):
+    '''
+    Matrix tricks to remove the mean from a sparse matrix.
```
Comment (Elarnon) [inline zero/svd2.py:11]: Should be more specific: this does not remove the mean, it removes the mean of the non-zero elements per row (I think — I'm not entirely sure how the `np.diff(sp_matrix.indptr)` works).

Also: if we are doing a copy of the `sp_matrix` anyways… I believe the following is equivalent and more readable:

```python
mask = csr_matrix((np.ones_like(sp_matrix.data), sp_matrix.indices, sp_matrix.indptr), shape=sp_matrix.shape)
row_counts = mask.sum(axis=1).A1
row_counts[row_counts == 0] = 1  #  `row_counts += row_counts == 0` if you feel evil.
means = row_sums / row_counts
return sp_matrix - diags(means) * mask, means
```

**Comment by Elarnon** [comment]
Nice!  Given the goal discussed in the previous meeting of making this code easier to use and reuse: it might make sense to document somewhere the distinction between `svd`, `svd2` and `ssvd` (which I thought was "sparse SVD" but apparently is "surprise SVD"?).  In particular, are `svd` and `svd2` solving the same mathematical problems through different means, or are there subtle distinctions?  When would one want to use the old `svd` instead of the new `svd2`?

File: zero/svd2.py (hunk 1 above)
Comment (jilljenn) [reply zero/svd2.py:11]: Hi, sorry I don't think the `indptr` is more readable.
So I think the `row_counts = mask.sum(axis=1).A1` is indeed better; so we remove every occurrence of `indices` and `indptr`.

**Comment by jilljenn** [comment]
Hi, thanks, fixed it. I started to prepare sphinx-apidoc (will be another PR).

- SVD2 was made to compare to SVD.
- SVD (in production) should stay there for some time while we make benchmarks on the prod.
- SVD is a randomized SVD during 10 iterations while SVD2 is (also) a randomized SVD but that converges to machine precision.
- SVD2 never computes the nb_users x nb_items but we don't know yet if the Kimsufi requires it.
- Surprise SVD was also there for benchmarking but the problem is that it requires Surprise (of course) which is a bit heavy for mangaki/zero.

File: zero/svd2.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    @property
+    def is_serializable(self):
+        """
+        Check whether we can save the model.
+        """
+        return True
+
+    def make_matrix(self, X, y):
+        """
+        Make a sparse matrix out of X and y.
```
Comment (Elarnon) [inline zero/svd2.py:53]: Would be even better to explain what `X`, `y`, `ratings` and `means` are (shape/meaning) :)  Maybe it is documented in `RecommendationAlgorithm`?

**Review by Elarnon — APPROVED** [review]

File: zero/svd2.py (hunk 2 above)
Comment (jilljenn) [reply zero/svd2.py:53]: フフフ of course it is not but thanks!

**Pull request merged**